types/artifacts.d.ts

```diff
@@ -307,7 +307,7 @@ declare global {
           files: Record<string, number>;
           unmappedBytes: number;
           totalBytes: number;
-        };
+        } | {error: string};
```

Comment:
open to some other format for this. ideas?

Reply:
> open to some other format for this. ideas?

Checking with `in` is kind of gross and ancient-feeling js, but it's also not the worst thing ever.

The other approach would be something like is done with `SourceMap` above this or [`NumericProduct`/`NonNumericProduct`](https://github.com/GoogleChrome/lighthouse/blob/478a954f85746b3b36cb09fc9132fcd870617973/types/audit.d.ts#L90-L104)

```ts
export interface Bundle {
  rawMap: RawSourceMap;
  script: ScriptElement;
  map: TextSourceMap;
  sizes: {
    files: Record<string, number>;
    unmappedBytes: number;
    totalBytes: number;
  };
  errorMessage?: never;
} | {
  rawMap: RawSourceMap;
  script: ScriptElement;
  map: TextSourceMap;
  errorMessage: string;
}
```
more verbose in the type definition, but here the `if (!bundle.errorMessage)  { /* we know bundle.sizes exists */ }` check is more idiomatic and is still required before tsc will let code access `bundle.sizes`.

Reply:
Belay that, [this doesn't even work for `NumericProduct`](https://github.com/GoogleChrome/lighthouse/blob/478a954f85746b3b36cb09fc9132fcd870617973/lighthouse-core/audits/audit.js#L290), but I think we can do better here. I'll play around with it, but per the above, the current approach works fine too.

Reply:
Ah yeah, you'd want `undefined`, not `never` (since `never` disappears in unions)
```ts
export type Bundle = {
  sizes: {
    // ...
  };
  errorMessage?: undefined;
} | {
  errorMessage: string;
}
```

_but_ because `errorMessage` is a string and `''` is falsy, you have to do a full check for undefined to discriminate:

```js
/** @type {LH.Artifacts.Bundle} */
const bundle;
if (bundle.errorMessage === undefined) {
  // we know bundle.sizes exists
}
```
which also feels kind of gross, so pick a flavor of awkwardness for this :)

Reply:
thanks for the alternative. if it's all the same, let's keep the current grossness

Reply:
> if it's all the same

technically the `in` check isn't type safe due to structural typing, but I don't see it being a problem in practice since the computed artifact is making the object